driver-examples/src/main/java/com/datastax/driver/examples/retry/DowngradingRetry.java

```diff
@@ -0,0 +1,470 @@
+ * feel free to adapt it to your application needs;</li>
+ * <li>You should never attempt to retry a non-idempotent write. See the driver's manual page on idempotence
+ * for more information.</li>
+ * </ul>
+ *
+ * @see <a href="http://datastax.github.io/java-driver/manual/">Java driver online manual</a>
+ */
+@SuppressWarnings("deprecation")
+public class DowngradingRetry {
+
+    private static final String[] CONTACT_POINTS = {"127.0.0.1"};
+
+    private static final int PORT = 9042;
+
+    /**
+     * The maximum number of retries to attempt.
+     */
+    private static final int MAX_RETRIES = 1;
+
+    /**
+     * The initial consistency level to use.
+     */
+    private static final ConsistencyLevel INITIAL_CL = QUORUM;
+
+    public static void main(String[] args) {
+
+        DowngradingRetry client = new DowngradingRetry(MAX_RETRIES);
+
+        try {
+
+            client.connect(CONTACT_POINTS, PORT);
+            client.createSchema();
+            client.write(INITIAL_CL, 0);
+            ResultSet rows = client.read(INITIAL_CL, 0);
+            client.display(rows);
+
+        } finally {
+            client.close();
+        }
+    }
+
+    private final int maxRetries;
+
+    private Cluster cluster;
+    private Session session;
+
+    private DowngradingRetry(int maxRetries) {
+        this.maxRetries = maxRetries;
+    }
+
+    /**
+     * Initiates a connection to the cluster specified by the given contact points and port.
+     *
+     * @param contactPoints the contact points to use.
+     * @param port          the port to use.
+     */
+    private void connect(String[] contactPoints, int port) {
+
+        cluster = Cluster.builder()
+                .addContactPoints(contactPoints).withPort(port)
+                .build();
+
+        System.out.println("Connected to cluster: " + cluster.getClusterName());
+
+        session = cluster.connect();
+    }
+
+    /**
+     * Creates the schema (keyspace) and table for this example.
+     */
+    private void createSchema() {
+
+        session.execute("CREATE KEYSPACE IF NOT EXISTS downgrading WITH replication " +
+                "= {'class':'SimpleStrategy', 'replication_factor':3}");
+
+        session.execute(
+                "CREATE TABLE IF NOT EXISTS downgrading.sensor_data (" +
+                        "sensor_id uuid," +
+                        "date date," + // emulates bucketing by day
+                        "timestamp timestamp," +
+                        "value double," +
+                        "PRIMARY KEY ((sensor_id,date),timestamp)" +
+                        ")");
+
+    }
+
+    /**
+     * Inserts data, retrying if necessary with a downgraded CL.
+     *
+     * @param cl         the consistency level to apply.
+     * @param retryCount the current retry count.
+     * @throws DriverException if the current consistency level cannot be downgraded.
+     */
+    private void write(ConsistencyLevel cl, int retryCount) {
```

Comment:
The use case you describe in the docs:

>  We recognize that there are use cases where downgrading is 
good – for instance, a dashboard application would present the latest information 
by reading at QUORUM, but it's acceptable for it to display stale information 
by reading at ONE sometimes. Users will now have to implement this logic directly
in their application, or alternatively, implement their own retry policy.

Wouldn't that only apply to reads?  Should we only show examples for application-level retries for reads?

Reply:
It's indeed imo harder to find valid examples of downgrading writes than it is for reads, but I think we should strive to give a complete example that works for both reads and writes.